src/python/WMCore/MicroService/Unified/MSRuleCleaner.py

```diff
@@ -528,18 +573,60 @@ def getMSOutputTransferInfo(self, wflow):
             wflow['TransferDone'] = True
         return wflow
 
-    def getRucioRules(self, wflow, gran, rucioAcct):
+    def setParentDatasets(self, wflow):
+        """
+        Used to resolves parent datasets for a workflow.
+        :param  wflow:   A MSRuleCleaner workflow representation
+        :return:         The workflow object
+        """
+        if wflow['InputDataset'] and wflow['IncludeParents']:
+            childDataset = wflow['InputDataset']
+            parentDataset = findParent(childDataset, self.msConfig['dbsUrl'])
+            if parentDataset:
+                wflow['ParentDataset'].append(parentDataset[childDataset[0]])
+                wflow['ParentDataset'] = list(set(wflow['ParentDataset']))
+                msg = "Found parent %s for input dataset %s for workflow: %s "
+                self.logger.debug(msg, parentDataset, wflow['InputDataset'], wflow['RequestName'])
+            else:
+                msg = "Could not find parent for input dataset: %s for workflows: %s"
+                self.logger.info(msg, wflow['InputDataset'], wflow['RequestName'])
+        return wflow
+
+    def getRucioRules(self, wflow, dataSpec, gran, rucioAcct):
```

Comment:
Just an idea here. How about we drop the usage of this `dataSpec` parameter and actually implement a map (could be in the `__init__` method, or even here) like:
```
mapRuleType = {"wmcore_output": ["OutputDatasets"],
                           "wmcore_transferor": ["InputDataset", "MCPileup", "DataPileup", "ParentDataset"]}
```

and deal with all input data rules in one execution of this method. 
This way, you do not need to call this same method with multiple parameters in the pipeline definition.

Reply:
Not a bad idea BTW. But I think the way it is implemented right now is more flexible, though. The map will always tie the behavior of that function to the account we are calling it with. And as we saw with this exact function in just this PR, there are indeed, cases when we need more freedom. One comes to mind immediately - splitting the RelVals under a different account and taking rule metadata into consideration.

Reply:
That's correct. However, the future relval use case will have a different behavior solely on the rucio account used. It's not related at all to the `dataSpec` parameter here. 
Besides, if we change and remove this `dataSpec` parameter, we only need to call this function once in the pipeline, making the code cleaner and less error prone.

Reply:
Ok @amaltaro I would like to drag your attention to one particular commit here. I implemented the so requested map in this commit [1]. Need to make a strong note here: I really do not think this is the correct imlementation.  And here is why:
   * This way the code looks really ugly and cumbersome to me.
   * The function complexity rose from `O(n)` to O(n^2) because of the nested for cycles. Need to mention though, this is just in theory, because in practice we expect for only one of the rucio accounts to have more than a single container list to search through, but nonetheless this is not an excuse in my opinion.
   * Now we have hard coded an entity such as an account as a string in the code, which is supposed to be only configuration dependent. And not only that but we tied a function behavior to it, meaning any new need for expanding the set of values of the configuration parameter (in this case the set of rucio accounts we consider into the service), we need to go into the code and make a relevant change of the so hard coded list. (an error I already experienced, while forgetting about the existence of `wma_test` account in parallel with the `wma_prod` account). This was absolutely not the case up to now. 
   * Which leads to another dependence we introduce here - dependence of the code itslef on the deployment variation (int or prod). 
   * There are a lot of peculiarities and specific custom cases that needed to be taken into account in a bunch of `ifs` which make it actually much more error prone than what it was before. The very proof of that  is the fact that while I had to modify it that way I have stumbled on at least 15 exceptions, all for different hidden under water stones like the one I already mentioned
   * Thoroughgoing this heavier logic inside the function instead of keeping it up in the set of function calls we make from the pipeline is exactly the opposite of the whole idea of constructing the service the way it is constructed at the first place. More precisely, to expose the logic as high as possible in the construction, while keeping the functions to be as minimalistic as possible and loaded with as little as possible additional logic.

So all that said, I would like to have your honest opinion, do you still like the variant with the account mapping inside the function more.

[1]
https://github.com/dmwm/WMCore/pull/10224/commits/d1b396017b1086c117b6f5ade72702637196ac3f

Reply:
> This way the code looks really ugly and cumbersome to me.
> The function complexity rose from O(n) to O(n^2) because of the nested for cycles. Need to mention though, this is just in theory, because in practice we expect for only one of the rucio accounts to have more than a single container list to search through, but nonetheless this is not an excuse in my opinion.

That's correct. However, the outer `for` loop is either a single or four elements, so in practice it's basically O(n). Now you only call this method once though, instead of the previous 3 calls with different `dataSpec`.

> Now we have hard coded an entity such as an account as a string in the code, which is supposed to be only configuration dependent. And not only that but we tied a function behavior to it, meaning any new need for expanding the set of values of the configuration parameter (in this case the set of rucio accounts we consider into the service), we need to go into the code and make a relevant change of the so hard coded list. (an error I already experienced, while forgetting about the existence of wma_test account in parallel with the wma_prod account). This was absolutely not the case up to now.

I don't think we need to hard code the account. I have made some changes on top of your latest commit such that you can see it. In addition to that, the `dataSpec` parameters were hard-coded as well... And if we need to consider a new rucio account, the source code has to change anyways, even before this PR was created.

> Which leads to another dependence we introduce here - dependence of the code itslef on the deployment variation (int or prod).

This should be automatically taken care of with the `self.msConfig` object.

> Thoroughgoing this heavier logic inside the function instead of keeping it up in the set of function calls we make from the pipeline is exactly the opposite of the whole idea of constructing the service the way it is constructed at the first place. More precisely, to expose the logic as high as possible in the construction, while keeping the functions to be as minimalistic as possible and loaded with as little as possible additional logic.

I can't really evaluate the tradeoff here between a) making many function calls with different arguments (expected arguments in the method) and executing the correct logic based on that; or b) making one function call with less arguments, and deciding the logic to be executed according an object attribute.
The logic though, is pretty small, the whole method source code is less than 40 lines (and only because you decided to deal with containers and blocks in different ways, otherwise it would be even smaller ;)).

As said above, I took your code (4th commit) and made some modifications on top of it, which I consider to simplify things a bit and might look better to you now:
```
...
        currPline = wflow['PlineMarkers'][-1]

        # Create the container list to the rucio account map and set the checkGlobalLocks flag.
        mapRuleType = {self.msConfig['rucioWmaAccount']: ["OutputDatasets"],
                       self.msConfig['rucioMStrAccount']: ["InputDataset", "MCPileup",
                                                           "DataPileup", "ParentDataset"]}
        if rucioAcct == self.msConfig['rucioMStrAccount']:
            checkGlobalLocks = True
        else:
            checkGlobalLocks = False

        # Find all the data placement rules created by the components:
        for dataType in mapRuleType[rucioAcct]:
            for dataCont in wflow[dataType]:
                self.logger.debug("getRucioRules: dataCont: %s", pformat(dataCont))
                if checkGlobalLocks and dataCont in self.globalLocks:
...
```

If you still think your initial implementation looks more robust, then please drop the 4th commit and I apologize for wasting your time.

Reply:
@amaltaro your code is breaking [2], just as I expected, because you are not taking into account the fact that not all of those data fields are lists any more. The `InputDataset` is a str. Notice that the code is trying to call rucio for every single letter in the dataset name. I was homogenizing (on puropuse) the structure of all those fields to lists in an earlier stage of this review, but you claimed this was bad [1]), I did the cahnge relentlesly, but from there on, we need to carry and remember that difference on the following steps. So I will have to think of a way out of this here at place.

[1]
https://github.com/dmwm/WMCore/pull/10224#discussion_r572978751

[2]

```
...
021-02-12 17:40:54,339:DEBUG:MSRuleCleaner:getRucioRules(): getRucioRules: dataCont: u'1'
2021-02-12 17:40:54,347:DEBUG:connectionpool:_make_request(): http://cmsrucio-int.cern.ch:80 "GET /rules/?scope=cms&account=wmcore_transferor&name=1 HTTP/1.1" 200 None
2021-02-12 17:40:54,348:DEBUG:MSRuleCleaner:getRucioRules(): getRucioRules: dataCont: u'/'
2021-02-12 17:40:54,359:DEBUG:connectionpool:_make_request(): http://cmsrucio-int.cern.ch:80 "GET /rules/?scope=cms&account=wmcore_transferor&name=%2F HTTP/1.1" 200 None
2021-02-12 17:40:54,360:DEBUG:MSRuleCleaner:getRucioRules(): getRucioRules: dataCont: u'R'
2021-02-12 17:40:54,369:DEBUG:connectionpool:_make_request(): http://cmsrucio-int.cern.ch:80 "GET /rules/?scope=cms&account=wmcore_transferor&name=R HTTP/1.1" 200 None
2021-02-12 17:40:54,370:DEBUG:MSRuleCleaner:getRucioRules(): getRucioRules: dataCont: u'E'
2021-02-12 17:40:54,378:DEBUG:connectionpool:_make_request(): http://cmsrucio-int.cern.ch:80 "GET /rules/?scope=cms&account=wmcore_transferor&name=E HTTP/1.1" 200 None
2021-02-12 17:40:54,379:DEBUG:MSRuleCleaner:getRucioRules(): getRucioRules: dataCont: u'C'
2021-02-12 17:40:54,389:DEBUG:connectionpool:_make_request(): http://cmsrucio-int.cern.ch:80 "GET /rules/?scope=cms&account=wmcore_transferor&name=C HTTP/1.1" 200 None
2021-02-12 17:40:54,389:DEBUG:MSRuleCleaner:getRucioRules(): getRucioRules: dataCont: u'O'
...

2021-02-12 17:43:54,496:ERROR:MSRuleCleaner:_dispatchWflow(): plineMSTrBlock: General error from pipeline. Workflow: tivanov_TaskChain_InclParents_HG2102_Val_210209_001757_1463. Error:  
An unknown exception occurred.
Details: no error information passed (http status code: 504 ('gateway_timeout',)). 
Will retry again in the next cycle.
Traceback (most recent call last):
  File "/afs/cern.ch/user/t/tivanov/work/private/WMCoreDev.d/WMCore/src/python/WMCore/MicroService/Unified/MSRuleCleaner.py", line 315, in _dispatchWflow
    pline.run(wflow)
  File "/afs/cern.ch/user/t/tivanov/work/private/WMCoreDev.d/WMCore/src/python/Utils/Pipeline.py", line 140, in run
    return reduce(lambda obj, functor: functor(obj), self.funcLine, obj)
  File "/afs/cern.ch/user/t/tivanov/work/private/WMCoreDev.d/WMCore/src/python/Utils/Pipeline.py", line 140, in <lambda>
    return reduce(lambda obj, functor: functor(obj), self.funcLine, obj)
  File "/afs/cern.ch/user/t/tivanov/work/private/WMCoreDev.d/WMCore/src/python/Utils/Pipeline.py", line 72, in __call__
    return self.run(obj)
  File "/afs/cern.ch/user/t/tivanov/work/private/WMCoreDev.d/WMCore/src/python/Utils/Pipeline.py", line 75, in run
    return self.func(obj, *self.args, **self.kwargs)
  File "/afs/cern.ch/user/t/tivanov/work/private/WMCoreDev.d/WMCore/src/python/WMCore/MicroService/Unified/MSRuleCleaner.py", line 642, in getRucioRules
    blocks = self.rucio.getBlocksInContainer(dataCont)
  File "/afs/cern.ch/user/t/tivanov/work/private/WMCoreDev.d/WMCore/src/python/WMCore/Services/Rucio/Rucio.py", line 229, in getBlocksInContainer
    if not self.isContainer(container):
  File "/afs/cern.ch/user/t/tivanov/work/private/WMCoreDev.d/WMCore/src/python/WMCore/Services/Rucio/Rucio.py", line 824, in isContainer
    response = self.cli.get_did(scope=scope, name=didName)
  File "/afs/cern.ch/work/t/tivanov/private/WMCoreDev.d/WMCore.venv/lib/python2.7/site-packages/rucio/client/didclient.py", line 394, in get_did
    raise exc_cls(exc_msg)
RucioException: An unknown exception occurred.
Details: no error information passed (http status code: 504 ('gateway_timeout',))
```

Reply:
Right! This logic considers the docSchema template:
https://github.com/dmwm/WMCore/pull/10224/files#diff-414c04814f1133dc334c7718010c11e70809389d23ec30e9d00d5e8874b84bf5R192

I guess you missed updating it to `str` over there then? Right now it's still meant to be of a `list` type.

Reply:
Ohh, I was looking at the docstring instead of the source code. Without reading the rest of the code, I don't know what is the expected data type for those new parameters. 
However, I think we define call InputDataset/ParentDataset as a string, MCPileup/DataPileup as a list (of strings).

Reply:
Hi @amaltaro, At least for ParentDataaset I would prefer to keep it as a list. It is already a list and is used as such in other places already too. Also, when we resolve the parent datasets from DBS the function `findParent()` returns a list. 

And I actually, do not see a reason why should we have some of the fields which we use to store dataset names in strings and some in lists. I know that those you suggest to be strings if we keep them as lists they can never become lists longer than a single element, but this I would say is not that bad. If we keep juggling with two different types of structures then in every function we have to take this into account, while if we convert them all to lists from the very beginning then the rest of the code simplifies extremely - we just aggregate the total list of datasets a function must work on, just by chaining all the already constructed lists. Or even better, we do not need to change a function behavior just because it was called to work on `wflow['InputDataset']` instead of `wflow['MCPileup']` or `wflow['OutputDatasets']`. The other way forces us to account for these data structure differences based on the type of input data every time we try to iterate through these workflow keys related to keeping dataset names.

Reply:
Hi @amaltaro I just force pushed your suggested code change plus a little modification, which we should apply every time if we keep the data structures indifferent forms. Please take a look again.  I am fine anyway.

Reply:
I think this  one is working and in the shape you wanted it. I am about to resolve this conversation.